rfc/0010-rancher-integration-policies.md

```diff
@@ -0,0 +1,188 @@
+Wasm module containing raw data, for us with the name
+`KUBEWARDEN_CUSTOM_SECTION_METADATA`). `rules` metadata is present there
+already.
+
+With this, annotated policies will contain almost all metadata needed.
+Information about if the policy is signed couldn't be included there by
+definition.
+
+Then, Rancher Explorer could make use of a client for OCI registries
+(such as [oci-registry-js](https://www.npmjs.com/package/oci-registry-js) or a
+wasm-compiled Rust library for example), to pull and cache each Wasm module.
+With [`WebAssembly.Module.customSections()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Module/customSections)
+it would be possible to extract all the metadata from the annotations of the
+policy.
+
+This would mean to pull hundreds of megabytes, as it will download the Wasm
+module of the policies, regardless of pagination and caching.
+
+Airgap installations would consist on mirroring the OCI registry (since metadata
+in stored in the Wasm modules).
+
+## B. Storing policy metadata in the OCI registry
+
+Same as alternative (A), but storing the policies' metadata in the OCI registry
+either under:
+- A new layer, with `config.mediaType` [creating an artifact type
+`application/vnd.oci.yaml-sample.config.v3+yaml`](https://github.com/opencontainers/artifacts/blob/main/artifact-authors.md#defining-a-unique-artifact-type).
+  This means pulling layers, and possibly uncompressing them.
+- Or in `manifest.config`
+  (https://github.com/opencontainers/artifacts/blob/main/artifact-authors.md#optional-defining-config-schema).
+  Which is not compressed, and whose checksum matches the checksum of the
+  uncompressed contents of the `manifest.config`.
+
+It would then be retrieved with an OCI registry client. This would be more
+efficient than pulling full Wasm modules for all policies as in (A).
+
+This means policies would have the metadata saved in both:
+- Wasm module, as of today. Used by `kwctl`, `policy-server`.
+- OCI registry `manifest.config`.
+
+This would mean that policy-related metadata can be dropped from
+`policy/artifacthub-pkg.yml`, which should only contain things related to
+Artifact Hub.
+
+Airgap installations would consist on mirroring the OCI registry.
+Rancher airgap works by providing a list of images to be mirrored, and
+instructions for using a private registry. We would need to instruct Kubewarden
+users to use an additional OCI private registry, and an OCI-capable client for
+performing the mirroring.
+
+This alternative provides the same implementation for both kubewarden-owner
+policies, Rancher-mirrored policies, and 3rd party policies. Just serve each of
+them on their respective OCI registries.
+
+A drawback could be finding frontend-capable OCI registry clients that can deal
+with these formats, and can authenticate against private registries too.
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
```

Comment:
(randomly starting thread here)

> In my opinion we should leave https://github.com/rancher/charts just for apps that can be deployed in your cluster (kubewarden for example)
>
> We could create a separate chart repository for the policies something like https://github.com/rancher/kubewarden-policies-charts. Drawback is that users would need to add it manually, or would it possible to add this new repository when kubewarden is installed? (I don't think that's possible, but I'm not sure..)

Submitting the Kubewarden policies into the Rancher distribution means that they will be supported in Rancher, which is what we want. They should all depend on the specific charts (ideally, just 1, that installs and configures the whole Kubewarden stack).

Reply:
You have a point there. Maybe the RFC will need to be amended in the future to submit the policies to a different repo. For now, I'm happy as it is, I'm more concerned with proving is viable by crafting the Helm charts and questions.yaml.

Reply:
yes, I agree. I'm happy as it is for now too :smile: